packages/core/echo/echo-schema/src/effect/dynamic/dynamic-schema.ts

```diff
@@ -0,0 +1,145 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import * as AST from '@effect/schema/AST';
+import * as S from '@effect/schema/Schema';
+
+import { invariant } from '@dxos/invariant';
+
+import { StoredEchoSchema } from './stored-schema';
+import { effectToJsonSchema, jsonToEffectSchema } from '../json-schema';
+import { type Identifiable } from '../reactive';
+import { schemaVariance } from '../utils';
+
+export interface DynamicSchemaConstructor extends S.Schema<DynamicEchoSchema> {
+  new (): Identifiable;
+}
+
+export const DynamicObjectSchemaBase = (): DynamicSchemaConstructor => {
+  return class {
+    static get ast() {
+      return StoredEchoSchema.ast;
+    }
+
+    static readonly [S.TypeId] = schemaVariance;
+    static get annotations() {
+      return StoredEchoSchema.annotations.bind(StoredEchoSchema);
+    }
+
+    static get pipe() {
+      return StoredEchoSchema.pipe.bind(StoredEchoSchema);
+    }
+  } as any;
+};
+
+export class DynamicEchoSchema extends DynamicObjectSchemaBase() implements S.Schema<any> {
+  private _schema: S.Schema<any> | undefined;
+  private _isDirty = true;
+  constructor(public readonly serializedSchema: StoredEchoSchema) {
+    super();
+  }
+
+  public override get id(): string {
+    return this.serializedSchema.id;
+  }
+
+  public get typename(): string {
+    return this.serializedSchema.typename;
+  }
+
+  public addColumns(columns: S.Struct.Fields) {
+    const oldSchema = this._getSchema();
+    const schemaExtension = S.partial(S.struct(columns));
+    const extended = S.extend(oldSchema, schemaExtension) as S.Schema<any>;
+    this.serializedSchema.jsonSchema = effectToJsonSchema(extended);
+  }
+
+  public updateColumns(columns: S.Struct.Fields) {
+    const oldAst = this._getSchema().ast;
+    invariant(AST.isTypeLiteral(oldAst));
+    const propertiesToUpdate = (S.partial(S.struct(columns)).ast as AST.TypeLiteral).propertySignatures;
+    const updatedProperties: AST.PropertySignature[] = [...oldAst.propertySignatures];
+    for (const property of propertiesToUpdate) {
+      const index = updatedProperties.findIndex((p) => p.name === property.name);
+      if (index >= 0) {
+        updatedProperties.splice(index, 1, property);
+      } else {
+        updatedProperties.push(property);
+      }
+    }
+    const newAst: any = { ...oldAst, propertySignatures: updatedProperties };
+    const schemaWithUpdatedColumns = S.make(newAst);
+    this.serializedSchema.jsonSchema = effectToJsonSchema(schemaWithUpdatedColumns);
+  }
```

Comment:
What happens when a column update causes existing entities to become invalid based on the new schema? Just curious if there are plans for how we're going to handle schema / data migrations.

Reply:
yeah, that's a tough question.
validations on write will be performed against the new schema, there are however currently no validations on read, so old values not matching the new type will be returned. for dynamic schema it would be more correct to return undefined in such case, since all columns are optional undefined will always match the schema. 
performing a validation on every read does sound like an undesirable overhead. but I think we'll need to get back to this topic at some point, in the context of lenses and different schema version interoperability